keps/sig-network/4969-cluster-domain-downward-api/README.md

```diff
@@ -0,0 +1,938 @@
+kubeadm[^prior-art-kubeadm], k3s[^prior-art-k3s], and OpenShift[^prior-art-openshift]).
+However, these are all inconsistent with each other, requiring applications to
+provide special cases for each.
+
+[^prior-art-kubeadm]: kubeadm (including kind) creates a ConfigMap `kube-system/kubeadm-config` that contains the full kubeadm config, including `networking.dnsDomain`.
+[^prior-art-k3s]: k3s creates a ConfigMap `kube-system/clusterdns` that contains it as the `.data.clusterDomain` field. See <https://github.com/k3s-io/k3s/pull/1785>.
+[^prior-art-openshift]: OpenShift defines a custom DNS CRD that contains it as the `.status.clusterDomain` field. See <https://docs.openshift.com/container-platform/4.17/networking/dns-operator.html#nw-dns-view_dns-operator>.
+
+It can also be retrieved from the kubelet's `/configz` endpoint, however this is
+[considered unstable](https://github.com/kubernetes/kubernetes/blob/9d967ff97332a024b8ae5ba89c83c239474f42fd/staging/src/k8s.io/component-base/configz/OWNERS#L3-L5).
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- Making it easier to use and generate FQDNs for internally-visible Services
+- Reducing the difference between Kubernetes distributions.
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+- Disallowing relative domain names.
+- Modifying DNS resolution.
+- Centralizing management of the cluster domain name setting.
+- Exposing all kubelet configuration.
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+Add a new field mapping into the Downward API mechanism, that lets containers retrieve the cluster domain name.
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+The Pod `foo` needs to access its sibling Service `bar` in the same namespace.
+It adds two `env` bindings:
```

Comment:
Would it be simpler to have a downward API binding to request the FQDN of a particular service (which would include both the namespace and the cluster domain name in a single value).

```yaml
env:
  - name: SERVICE1
    valueFrom:
      serviceReference:
        name: bar
  - name: SERVICE2
    valueFrom:
      serviceReference:
        namespace: otherns
        name: blah
```

→

```sh
SERVICE1=bar.myns.svc.cluster.local
SERVICE2=blah.otherns.svc.cluster.local
```

Reply:
IMO tricky without ReferenceGrant. Users might expect that the service reference actually honors the existence of the other Service, or might make API changes to built that expectation.

Reply:
That also wouldn't really work for our (@stackabletech) use case/current API contract.

Our operators generate configmaps with all the details you'll need to connect to the service managed by the operators (including URLs). Our operators' pod manifests don't know what specific CR objects they'll end up managing.

Reply:
In that case, it would be good to update the user story (or add a second user story) explaining that. The user stories are there to help explain why the particular solution you chose is the best solution.

Reply:
We want to reduce DNS traffic (as you mentioned in https://github.com/kubernetes/enhancements/pull/4972#issuecomment-2651973659), and avoid hostname conflicts. Arguably the latter is the bigger problem.

`$service.$namespace` can match a public DNS domain if `$namespace` is set to an ICANN domain. Even though search takes priority so requests from the cluster would still get directed to the cluster-internal resource rather than the public website, that means that we're still issuing internal TLS certificates that impersonate public DNS domains.